contracts/Tokens/Prime/PrimeLiquidityProvider.sol

```diff
@@ -0,0 +1,209 @@
+import { AccessControlledV8 } from "@venusprotocol/governance-contracts/contracts/Governance/AccessControlledV8.sol";
+
+contract PrimeLiquidityProvider is Ownable2StepUpgradeable, AccessControlledV8 {
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    /// @notice The initial TOKEN index
+    uint224 public constant INITIAL_INDEX = 1e36;
+
+    /// @notice Address of the Prime contract
+    address public prime;
+
+    /// @notice The rate at which token is distributed (per block)
+    mapping(address => uint256) public tokenDistributionSpeeds;
+
+    /// @notice The rate at which token is distributed to the Prime contract
+    mapping(address => uint256) public lastAccruedBlock;
+
+    /// @notice The token accrued but not yet transferred to prime contract
+    mapping(address => uint256) public tokenAmountAccrued;
+
+    /// @dev This empty reserved space is put in place to allow future versions to add new
+    /// variables without shifting down storage in the inheritance chain.
+    uint256[46] private __gap;
+
+    /// @notice Emitted when a token distribution is initialized
+    event TokenDistributionInitialized(address indexed token);
+
+    /// @notice Emitted when a new token distribution speed is set
+    event TokenDistributionSpeedUpdated(address indexed token, uint256 newSpeed);
+
+    /// @notice Emitted when distribution state(Index and block) is updated
+    event TokensAccrued(address indexed token);
+
+    /// @notice Emitted when token is transferred to the prime contract
+    event TokenTransferredToPrime(address indexed token, uint256 amount);
+
+    /// @notice Thrown when arguments are passed are invalid
+    error InvalidArguments();
+
+    /**
+     * @param prime_ Address of the Prime contract
+     */
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(address prime_) {
+        prime = prime_;
+
+        _disableInitializers();
+    }
+
+    /**
+     * @notice Accrue token by updating the distribution state
+     * @param token Address of the token
+     * @custom:event Emits TokensAccrued event
+     */
+    function accrueTokens(address token) public {
+        uint256 distributionSpeed = tokenDistributionSpeeds[token];
+        uint256 blockNumber = getBlockNumber();
+
+        uint256 deltaBlocks = blockNumber - lastAccruedBlock[token];
+
+        if (deltaBlocks > 0 && distributionSpeed > 0) {
+            uint256 balance = IERC20Upgradeable(token).balanceOf(address(this));
+            uint256 accruedSinceUpdate = deltaBlocks * distributionSpeed;
+            uint256 tokenAccrued = (balance * accruedSinceUpdate);
+
+            lastAccruedBlock[token] = blockNumber;
+            tokenAmountAccrued[token] += tokenAccrued;
+        } else if (deltaBlocks > 0) {
+            lastAccruedBlock[token] = blockNumber;
+        }
+
+        emit TokensAccrued(token);
```

Comment:
We could add the amount of tokens accrued in this execution and the amount of pending tokens to be distributed after this execution

Reply:
Fixed

Reply:
Fixe, 4c4f9e47d569e3d4d8bae2177b4461f8bb870396